internal/resources/free_tier_cluster.go

```
@@ -0,0 +1,491 @@
+	_, err = f.Client.ExecuteWithRetry(
+		ctx,
+		cfg,
+		nil,
+		f.Token,
+		nil,
+	)
+	if err != nil {
+		resourceNotFound, errString := api.CheckResourceNotFoundError(err)
+		if resourceNotFound {
+			tflog.Info(ctx, "resource doesn't exist in remote server removing resource from state file")
+			response.State.RemoveResource(ctx)
+			return
+		}
+		response.Diagnostics.AddError(
+			"Error Deleting Free Tier Cluster",
+			"Could not delete cluster id "+state.Id.String()+": "+errString,
+		)
+		return
+	}
+
+	err = f.checkFreeTierClusterStatus(ctx, state.OrganizationId.ValueString(), state.ProjectId.ValueString(), state.Id.ValueString())
+	if err != nil {
+		resourceNotFound, errString := api.CheckResourceNotFoundError(err)
+		if !resourceNotFound {
+			response.Diagnostics.AddError(
+				"Error Deleting Capella Cluster",
+				"Could not delete cluster id "+state.Id.String()+": "+errString,
+			)
+			return
+		}
+		// resourceNotFound as expected.
+		return
+	}
+
+	// This case will only occur when cluster deletion has failed,
+	// and the cluster record still exists in the cp metadata. Therefore,
+	// no error will be returned when performing a GET call.
+	cluster, err := f.retrieveFreeTierCluster(ctx, state.OrganizationId.ValueString(), state.ProjectId.ValueString(), state.Id.ValueString())
+	if err != nil {
+		resourceNotFound, errString := api.CheckResourceNotFoundError(err)
+		if resourceNotFound {
+			tflog.Info(ctx, "resource doesn't exist in remote server removing resource from state file")
+			response.State.RemoveResource(ctx)
+			return
+		}
+		response.Diagnostics.AddError(
+			"Error Deleting Capella Cluster",
+			"Could not delete cluster id "+state.Id.String()+": "+errString,
+		)
+		return
+	}
+	response.Diagnostics.AddError(
+		"Error deleting cluster",
+		fmt.Sprintf("Could not delete cluster id %s, as current Cluster state: %s", state.Id.String(), cluster.CurrentState),
+	)
+}
```

this is looks over complicated.  just keep waiting for 404.  otherwise error out.

---

Hey @l0n3star , if the cluster goes in destroy failed state, we can throw the error notifying the state with this code and the same is  being done on the cluster side as well, I can change it check for 404 status but I feel existing one is covering an additional edge case

---

if cluster is in destroy failed state, then you won't 404.

---

checking for 404 would also catch edge case of destroy failed, as in that case you'd not get 404.  what happens on the backend is it checks the database for cluster record.  when cluster is successfully deleted, cluster record is deleted.  if destroy fails for some reason, cluster record is kept.

---

but in that case we need to keep polling until time out, if cluster is in destroyed state we would still be polling until timeout is reached as we do not get 404 status, I'd still feel existing approach is good though bit complicated to follow up on first glance, and another point is again it takes some effort to work on 404 approach as well, and finally existing is working and don't want to modify it as well ( because of some testing done already). :P :)

---

I've modified the code to check for cluster destroyed or not, if not I'm checking for the cluster is in destroyed failed state and throwing the error appropriately.